tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -0,0 +1,1414 @@
+   auto elementSize = element->GetSize();
+
+   auto pageSize = pageInfo.fLocator.fBytesOnStorage;
+   void *pageBuffer = malloc(std::max(pageSize, static_cast<std::uint32_t>(elementSize * pageInfo.fNElements)));
+   R__ASSERT(pageBuffer);
+   Read(pageBuffer, pageSize, pageInfo.fLocator.fPosition);
+
+   auto bytesOnStorage = (element->GetBitsOnStorage() * pageInfo.fNElements + 7) / 8;
+   if (pageSize != bytesOnStorage) {
+      fDecompressor(pageBuffer, pageSize, bytesOnStorage);
+      pageSize = bytesOnStorage;
+   }
+
+   if (!element->IsMappable()) {
+      pageSize = elementSize * pageInfo.fNElements;
+      auto unpackedBuffer = reinterpret_cast<unsigned char *>(malloc(pageSize));
+      R__ASSERT(unpackedBuffer != nullptr);
+      element->Unpack(unpackedBuffer, pageBuffer, pageInfo.fNElements);
+      free(pageBuffer);
+      pageBuffer = unpackedBuffer;
+   }
+
+   auto indexOffset = clusterDescriptor.GetColumnRange(columnId).fFirstElementIndex;
+   auto newPage = fPageAllocator->NewPage(columnId, pageBuffer, elementSize, pageInfo.fNElements);
+   newPage.SetWindow(indexOffset + firstInPage, RPage::RClusterInfo(clusterId, indexOffset));
+   fPagePool->RegisterPage(newPage,
+      RPageDeleter([](const RPage &page, void */*userData*/)
+      {
+         RPageAllocatorFile::DeletePage(page);
+      }, nullptr));
+   return newPage;
+}
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageSourceFile::PopulatePage(
+   ColumnHandle_t columnHandle, NTupleSize_t globalIndex)
+{
+   auto columnId = columnHandle.fId;
+   auto cachedPage = fPagePool->GetPage(columnId, globalIndex);
+   if (!cachedPage.IsNull())
+      return cachedPage;
+
+   auto clusterId = fDescriptor.FindClusterId(columnId, globalIndex);
+   R__ASSERT(clusterId != kInvalidDescriptorId);
+   const auto &clusterDescriptor = fDescriptor.GetClusterDescriptor(clusterId);
+   auto selfOffset = clusterDescriptor.GetColumnRange(columnId).fFirstElementIndex;
+   R__ASSERT(selfOffset <= globalIndex);
+   return PopulatePageFromCluster(columnHandle, clusterDescriptor, globalIndex - selfOffset);
+}
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageSourceFile::PopulatePage(
+   ColumnHandle_t columnHandle, const RClusterIndex &clusterIndex)
+{
+   auto clusterId = clusterIndex.GetClusterId();
+   auto index = clusterIndex.GetIndex();
+   auto columnId = columnHandle.fId;
+   auto cachedPage = fPagePool->GetPage(columnId, clusterIndex);
+   if (!cachedPage.IsNull())
+      return cachedPage;
+
+   R__ASSERT(clusterId != kInvalidDescriptorId);
+   const auto &clusterDescriptor = fDescriptor.GetClusterDescriptor(clusterId);
+   return PopulatePageFromCluster(columnHandle, clusterDescriptor, index);
+}
+
+void ROOT::Experimental::Detail::RPageSourceFile::ReleasePage(RPage &page)
+{
+   fPagePool->ReturnPage(page);
+}
+
+std::unique_ptr<ROOT::Experimental::Detail::RPageSource> ROOT::Experimental::Detail::RPageSourceFile::Clone() const
```

Comment:
You may want to specific in the doc string what this is meant to clone (data, meta, file descriptor, etc.)?